libs/stokesian_dynamics/device_matrix.hpp

```diff
@@ -0,0 +1,658 @@
+// Copyright (c) 2018 Henri Menke. All rights reserved.
+
+#include <iomanip>
+#include <limits>
+#include <type_traits>
+
+#include "thrust_wrapper.hpp"
+
+#ifdef __CUDACC__
+#include <cublas_v2.h>
+#include <cusolverDn.h>
+#endif
+
+#ifdef __CUDACC__
+#define DEVICE_FUNC __host__ __device__
+#else
+#define DEVICE_FUNC
+#endif
+
+#ifdef __GNUC__
+#define MAYBE_UNUSED __attribute__((unused))
+#else
+#define MAYBE_UNUSED
+#endif
+
+namespace policy {
+
+struct host {
+    template <typename T>
+    using vector = thrust_wrapper::host_vector<T>;
+    // the remove_const is necessary, because things cannot be static and const
+    // at the same time
+    static std::remove_const<decltype(thrust_wrapper::host)>::type par()
+      { return thrust_wrapper::host; }
+};
+
+struct device {
+    template <typename T>
+    using vector = thrust_wrapper::device_vector<T>;
+    static std::remove_const<decltype(thrust_wrapper::device)>::type par()
+      { return thrust_wrapper::device; }
+};
+
+template <typename...>
+using void_t = void;
+
+template <typename Policy, typename = void>
+struct is_policy : std::false_type {};
+
+template <typename Policy>
+struct is_policy<Policy, void_t<typename Policy::template vector<class T>,
+                                decltype(Policy::par())>> : std::true_type {};
+
+} // namespace policy
+
+DEVICE_FUNC inline thrust_wrapper::tuple<std::size_t, std::size_t>
+unravel_index(std::size_t index, std::size_t lda) {
+    std::size_t i, j;
+    i = index % lda;
+    index /= lda;
+    j = index;
+    return thrust_wrapper::make_tuple(i, j);
+    ;
+}
+
+/// \cond
+
+// LAPACK prototypes
```

Comment:
Why are those needed?

Reply:
Do you mean the LAPACK prototypes, or the comment lines?

Reply:
LAPACK traditionally has no header file, so you need these prototypes here.

Reply:
LAPACK routines are the alternatives to cuSOLVER routines, when solving on the CPU